demos/portal-embed-demo/public/js/portalog/portals-controller.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Portals Controller
+ * Controlling all the portal related features
+ */
+class PortalsController {
+
+    /**
+     * Initiating paratmeter
+     * @constructor
+     * @param {HTMLElement} root - the root element
+     * @param {String} origin - origin of portal
+     * @param {String} path - path of portal
+     * @param {Number} anmiateY - A target position to animate in px
+     */
+    constructor(root, origin, path, anmiateY) {
+        this.root = root;
+        this.origin = origin;
+        this.path = path;
+        this.anmiateY = anmiateY;
+    }
+
+    /**
+     * Embedding portal element to the root.
+     * @param {HTMLPortalElement} predecessor - optional
+     */
+    embed(predecessor) {
+
+        // clean up the root
+        this._remove(this.root.querySelector('portal'));
+
+        if (predecessor) {
+            // If the optional predecessor parameter was set
+            // add the predecessor to the root
+            this.portal = predecessor;
+        } else {
+            // Or else, create new portal
+            this.portal = document.createElement('portal');
+            this.portal.src = this.origin + this.path;
+            // https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
+            // TODO: autoplay workaround
+        }
+        const messageUI = this._genPortalMessageUI();
+        this.root.appendChild(this.portal);
+        this.root.appendChild(messageUI);
+
+        // Add animation on-click
+        this.portal.addEventListener('click', (evt) => {
+            this.initialY = this.root.getBoundingClientRect().y;
+            this.initialWidth = this.root.getBoundingClientRect().width;
+            this._remove(this.root.querySelector('.message-controller'));
+            this.root.style.transition =
+                `top 0.6s cubic-bezier(.49,.86,.37,1.01),
+                 left 0.3s cubic-bezier(.49,.86,.37,1.01), 
+                width 0.3s cubic-bezier(.49,.86,.37,1.01),
+                padding-top 0.3s cubic-bezier(.49,.86,.37,1.01)`;
+            this.root.style.top = `${this.anmiateY - this.initialY}px`;
+            this.root.style.width = '95%';
+            this.root.style.paddingTop = 'calc(95% * 0.65)';
+            this.root.style.left = 'calc((100% - 95%)/2)';
+            this.portal.postMessage({ control: 'hide' }, this.origin);
+        });
+
+    }
+
+    /**
+     * Adding event listeners
```

Comment:
This description is kind of vague. It might be worth just inlining this into the function that creates the DOM (if you take my suggestion above and split the adoptPredecessor case out to just `replaceWith`. At that point, actually, maybe both just ought to be in the `PortalsController` constructor, since you always want to call it after constructing that class.

Reply:
Agreed. I moved it to the constructor. (I don't remember why I designed it as a public function... so probably unnecessary stuff)